Code Annotation Docs (#579)
quarto-dev/quarto-web: docs/prerelease/1.3/code-annotation.qmd

---
title: Code Annotation
code-annotations: below
---

## Overview

Code blocks and executable code cells in Quarto may now include line-based annotations. Line-based annotations provide a way to attach explanation to lines of code much like footnotes. 

For example, this code uses annotation to describe the steps in an R dplyr pipeline in plain language:

```r
library(tidyverse)
library(palmerpenguins)
penguins |>                                            # <1>
  mutate(                                              # <2>
    bill_ratio = bill_depth_mm / bill_length_mm,       # <2>

>     bill_area  = bill_depth_mm * bill_length_mm        # <2>

Reviewer: Is this how one would calculate the area? Not that it's too important, but I'm guessing it's not 100% accurate since the bill isn't flat. No problem leaving it in though.

Author: Yeah, it's definitely a **rough** estimate. But, interestingly enough to provide pretty good (linear) separation of the species in a plot of ratio vs. area.

  )                                                    # <2>
```
1. Take `penguins`, and then,
2. add new columns for the bill ratio and bill area.

The default HTML annotation style displays annotations in a list below the code block. Clicking on the annotation number in the list highlights the relevant lines in the code. Other [HTML styles](#html-styles) hide the annotations, revealing them in a tooltip when a user hovers or selects a marker.

> The PDF format also numbers annotations and displays the annotation text below the code. In other formats, rather than numbering annotations, annotations are labelled with the line of code (or lines of code) to which the annotation text applies.

Reviewer: ```suggestion
The PDF format also allows for annotations, numbering, and displaying the annotation text below the code. In other formats, annotations are instead labeled with the line of code (or lines of code) to which the annotation text applies.
```

Reviewer: Would be good to mention one or two examples of "other formats", like Word, GitHub Markdown, ...?

::: panel-tabset
#### PDF

> ![](images/annote-pdf.png)

Author: Add ALT tag

#### GitHub Flavored Markdown

````markdown
``` r
library(tidyverse)
library(palmerpenguins)
penguins |>
  mutate(
    bill_ratio = bill_depth_mm / bill_length_mm,
    bill_area  = bill_depth_mm * bill_length_mm
  )
```

Line 3  
Take `penguins`, and then,

Lines 4-7  
add new columns for the bill ratio and bill area.
````
:::

> To add code annotation to a code block, you need to add two things: specially formatted code comments to your code, and an ordered list below your code with your annotation content. Read more in [Annotation Syntax](#annotation-syntax).

Reviewer: ```suggestion
To add code annotation to a code block, you need to add two things: specially formatted code comments in your code cell, and an ordered list below the code cell with the annotation text. Read more in [Annotation Syntax](#annotation-syntax).
```

The `code-annotations` option controls how annotations appear in the HTML format (`below` (default), `hover` or `select`), and in all formats, whether annotation is disabled (`false`), or if annotations should be removed from the output (`none`).

## Annotation Syntax {#annotation-syntax}

Annotations for a code cell consist of two related elements:

1.  Each annotated line should be terminated with a comment (using the code cell's language comment character) followed by a space and then an annotation number enclosed in angle brackets (e.g. `# <2>`). You may repeat an annotation number if the annotation spans multiple lines.

> 2.  An ordered list that appears immediately after the code cell which includes the contents of each annotation. Each numbered item in the ordered list will correspond with the line(s) of code with the same annotation number.

Reviewer: ```suggestion
2.  An ordered list that appears immediately after the code cell which includes the contents of each annotation. Each numbered item in the ordered list will correspond with the line(s) of code to the same annotation number.
```

For example, the annotations in the overview were produced with the following:

> ```` markdown

Reviewer: I usually use `echo: fenced` to show code chunks with the fences, is there a reason not to do that here?

Author: Because I also need to show the markdown ordered list below the code chunk.

```r
library(tidyverse)
library(palmerpenguins)
penguins |>                                      # <1>
  mutate(                                        # <2>
    bill_ratio = bill_depth_mm / bill_length_mm, # <2>
    bill_area  = bill_depth_mm * bill_length_mm  # <2>
  )                                              # <2>
```
1. Take `penguins`, and then,
2. add new columns for the bill ratio and bill area.
````

## Annotation Style {#html-styles}

For HTML output, there are three annotation styles you can set with the `code-annotations` document option:

`below`

:   By default (or if `code-annotations: below` is specified), code annotation text will appear below the code cell.

`hover`

:   Code annotation text will be displayed when the user hovers over the annotation marker for a line of code. 

>     ![](images/annote-hover.png){width=90%}

Author: Add ALT tag
